util/argo/argo.go

```diff
@@ -412,9 +412,18 @@ func GetAppProjectByName(name string, projLister applicationsv1.AppProjectLister
 	return GetAppVirtualProject(project, projLister, settingsManager)
 }
 
-// GetAppProject returns a project from an application
-func GetAppProject(spec *argoappv1.ApplicationSpec, projLister applicationsv1.AppProjectLister, ns string, settingsManager *settings.SettingsManager, db db.ArgoDB, ctx context.Context) (*argoappv1.AppProject, error) {
-	return GetAppProjectByName(spec.GetProject(), projLister, ns, settingsManager, db, ctx)
+// GetAppProject returns a project from an application. It will also ensure
+// that the application is allowed to use the project.
+func GetAppProject(app *argoappv1.Application, projLister applicationsv1.AppProjectLister, ns string, settingsManager *settings.SettingsManager, db db.ArgoDB, ctx context.Context) (*argoappv1.AppProject, error) {
+	proj, err := GetAppProjectByName(app.Spec.GetProject(), projLister, ns, settingsManager, db, ctx)
+	if err != nil {
+		return nil, err
+	}
+	if !proj.IsAppNamespacePermitted(app, ns) {
+		return nil, fmt.Errorf("application '%s' in namespace '%s' is not allowed to use project '%s'",
```

Comment:
Maybe add ", because the project does not allow applications in that namespace" or something to clarify why it's disallowed.